File: yjs-worker/src/types/internal_frame.ts
import { InternalFrameType } from "./internal_frame_type.js";
import { InternalFrameHeaderSize } from "../constants/header_size.js";
import { InternalChannelType } from "../types/internal_channel_type.js";
import { convertBytesToUUIDString, convertUUIDToBytes } from "../util/uuid.js";

export type InternalFrame = {
  version: number;
  type: InternalFrameType;
  connectionId: string;
  connectorChannelId: number;
  blockPackId: string;
  payload: Buffer;
};

export function parseInternalFrame(payload: Buffer): InternalFrame | null {
  // [version:1][type:1][channelType:1][connectionId:16][connectorChannelId:4][channelId:16][raw payload:n]
  if (payload.length < InternalFrameHeaderSize) {
    return null;
  }
  if (payload[2] !== InternalChannelType.InternalChannelType_BlockPack) {
    return null;
  }

  const type = payload[1] as InternalFrameType;
  if (!Object.values(InternalFrameType).includes(type)) {
    return null;
  }

  const connectorChannelId = payload.readUInt32BE(19);
  if (connectorChannelId === 0) {
    return null;
  }

  const connectionId = convertBytesToUUIDString(payload.subarray(3, 19));
  const blockPackId = convertBytesToUUIDString(payload.subarray(23, 39));
  if (connectionId === null || blockPackId === null) {
    return null;
  }

  return {
    version: payload[0],
    type,
    connectionId,
    connectorChannelId,
    blockPackId,
    payload: payload.subarray(InternalFrameHeaderSize),
  };
}

export function createInternalFrame(
  type: InternalFrameType,
  connectionId: string,
  connectorChannelId: number,
  blockPackId: string,
  payload: Buffer = Buffer.alloc(0),
): Buffer {
  const frame = Buffer.alloc(InternalFrameHeaderSize + payload.length);

  frame[0] = 1;
  frame[1] = type;
  frame[2] = InternalChannelType.InternalChannelType_BlockPack;
  convertUUIDToBytes(connectionId).copy(frame, 3);
  frame.writeUInt32BE(connectorChannelId, 19);
  convertUUIDToBytes(blockPackId).copy(frame, 23);
  payload.copy(frame, InternalFrameHeaderSize);

  return frame;
}

File: app/realtime/types/internal_frame.go
package realtimetypes

import (
	"encoding/binary"
	"errors"

	"github.com/google/uuid"

	constants "github.com/HiIamJeff67/notezy-backend/shared/constants"
)

type InternalFrame struct {
	Version            byte
	Type               InternalFrameType
	ChannelType        ChannelType
	ConnectionId       uuid.UUID
	ConnectorChannelId uint32
	ChannelId          uuid.UUID
	Payload            []byte
}

func (f *InternalFrame) UnmarshalBytes(payload []byte) error {
	*f = InternalFrame{}

	// [version:1][type:1][channelType:1][connectionId:16][connectorChannelId:4][channelId:16][raw payload:n]
	if len(payload) < constants.RealtimeInternalFrameHeaderSize {
		return errors.New("invalid realtime internal frame")
	}

	*f = InternalFrame{
		Version:            payload[0],
		Type:               InternalFrameType(payload[1]),
		ConnectorChannelId: binary.BigEndian.Uint32(payload[19:23]),
		Payload:            payload[39:],
	}
	switch InternalChannelType(payload[2]) {
	case InternalChannelType_BlockPack:
		f.ChannelType = ChannelType_BlockPack
	default:
		*f = InternalFrame{}

		return errors.New("invalid realtime internal frame")
	}

	copy(f.ConnectionId[:], payload[3:19])
	copy(f.ChannelId[:], payload[23:39])
	if f.ConnectionId == uuid.Nil || f.ConnectorChannelId == 0 || f.ChannelId == uuid.Nil {
		*f = InternalFrame{}

		return errors.New("invalid realtime internal frame")
	}

	return nil
}

func (f InternalFrame) MarshalBytes() ([]byte, error) {
	var internalChannelType InternalChannelType
	switch f.ChannelType {
	case ChannelType_BlockPack:
		internalChannelType = InternalChannelType_BlockPack
	default:
		return nil, errors.New("invalid realtime internal frame")
	}
	if f.ConnectionId == uuid.Nil || f.ConnectorChannelId == 0 || f.ChannelId == uuid.Nil {
		return nil, errors.New("invalid realtime internal frame")
	}

	payload := make([]byte, constants.RealtimeInternalFrameHeaderSize+len(f.Payload))
	payload[0] = f.Version
	payload[1] = byte(f.Type)
	payload[2] = byte(internalChannelType)
	copy(payload[3:19], f.ConnectionId[:])
	binary.BigEndian.PutUint32(payload[19:23], f.ConnectorChannelId)
	copy(payload[23:39], f.ChannelId[:])
	copy(payload[39:], f.Payload)

	return payload, nil
}
